### PR TITLE
[TOPI] Add support for groupped conv3d

### PR DESCRIPTION
**Pull request by tkonolige**
Change conv3d to use generic conv implementation which supports grouped convolutions. Also, remove support for non-float16 tensorcore operations as they cause large degradation in accuracy. Generic conv now supports autoscheduler.

@mbrookhart @jwfromm

### CODE REVIEW

File: src/target/source/codegen_cuda.cc [hunk 1]
```diff
@@ -1051,18 +1051,27 @@ void CodeGenCUDA::PrintWmmaScope(const std::string& scope, DataType t, const Var
   }
 }
 
+int stoi(const std::string& str) {
+  try {
+    return std::stoi(str);
+  } catch (std::invalid_argument& e) {
+    LOG(FATAL) << "Cannot convert \"" << str << "\" to int";
+    throw;
```
Comment (mbrookhart) [inline src/target/source/codegen_cuda.cc:1059]: I don't think you need this throw, LOG(FATAL) throws under the hood.

File: src/arith/rewrite_simplify.cc [hunk 2]
```diff
@@ -820,6 +820,10 @@ PrimExpr RewriteSimplifier::Impl::VisitExpr_(const FloorDivNode* op) {
     TVM_TRY_REWRITE_IF(floordiv(x * y, y), x, CanProveGreaterEqual(y.Eval(), 0));
     TVM_TRY_REWRITE_IF(floordiv(y * x, y), x, CanProveGreaterEqual(y.Eval(), 0));
 
+    if ((floordiv(x, y)).Match(ret) && analyzer_->CanProve(x.Eval() < y.Eval())) {
+      return 0;
+    }
+
```
Comment (mbrookhart) [inline src/arith/rewrite_simplify.cc:826]: I'm not sure what this is for?

**Review by mbrookhart — COMMENTED** [review]
A couple of nits. I'd love it if @masahi or @jwfromm could take a look re: tensorcores, but otherwise, I'm happy.

**Comment by masahi** [comment]
> Also, remove support for non-float16 tensorcore operations as they cause large degradation in accuracy

Yeah, I also noticed this implicit downcasting before, I'm in favor of this change. Users can always explicitly cast their inputs to fp16. 

This may break existing flows if some users depend on this behavior, which is also the default in cudnn/cublas. Personally I don't mind it, but if that's concerning we can always add a workaround to allow the old behavior.

File: python/tvm/relay/op/strategy/cuda.py [hunk 3]
```diff
@@ -641,7 +642,7 @@ def conv3d_strategy_cuda(attrs, inputs, out_type, target):
                     (N % 16 == 0 and CI % 16 == 0 and CO % 16 == 0)
                     or (N % 8 == 0 and CI % 16 == 0 and CO % 32 == 0)
                     or (N % 32 == 0 and CI % 16 == 0 and CO % 8 == 0)
-                ):
+                ) and out_type == "float16":
```
Comment (masahi) [inline python/tvm/relay/op/strategy/cuda.py:645]: How about making this change to conv2d as well?. Perhaps better to send a separate PR for disabling fp32 tensorcore use since that would be a breaking change and potentially controversial.

File: src/arith/rewrite_simplify.cc (hunk 2 above)
Comment (mbrookhart) [reply src/arith/rewrite_simplify.cc:826]: Maybe need to a a pass test for this.

Comment (tkonolige) [reply src/arith/rewrite_simplify.cc:826]: This is needed so that the simplifier here: https://github.com/apache/tvm/pull/9873/files#diff-04a1aab966320b6f63c390cc8b79f79b543a1a7a3f560ea3a97d18c2bb041a0aR837-R839 will remove the division by groups. Without this, simplification get stuck simplifying `ff // (num_filter // groups) * (in_channel // groups) + rc` to `ff // num_filter * in_channel + rc`. It can't prove that `ff // num_filter = 0`.

File: src/target/source/codegen_cuda.cc (hunk 1 above)
Comment (tkonolige) [reply src/target/source/codegen_cuda.cc:1059]: Somehow the compiler can't see through LOG(FATAL). I think it is because it actually throws in the destructor.

File: python/tvm/relay/op/strategy/cuda.py (hunk 3 above)
Comment (tkonolige) [reply python/tvm/relay/op/strategy/cuda.py:679]: This was causing an issue with the new schedule, so I disabled it.

**Comment by tkonolige** [comment]
@masahi @mbrookhart I've finally got this branch green. Could you re-review?

**Review by mbrookhart — APPROVED** [review]
LGTM. I would love to align on the tensorcore issues across the ops, if any still remain, but that can be a second PR.

**Review by masahi — APPROVED** [review]
